pkg/kubectl/metricsutil/metrics_client.go

```diff
@@ -0,0 +1,135 @@
+	DefaultHeapsterService   = "heapster"
+	DefaultHeapsterPort      = "" // use the first exposed port on the service
+)
+
+type HeapsterMetricsClient struct {
+	Client            *client.Client
+	HeapsterNamespace string
+	HeapsterScheme    string
+	HeapsterService   string
+	HeapsterPort      string
+}
+
+func NewHeapsterMetricsClient(client *client.Client, namespace, scheme, service, port string) *HeapsterMetricsClient {
+	return &HeapsterMetricsClient{
+		Client:            client,
+		HeapsterNamespace: namespace,
+		HeapsterScheme:    scheme,
+		HeapsterService:   service,
+		HeapsterPort:      port,
+	}
+}
+
+func DefaultHeapsterMetricsClient(client *client.Client) *HeapsterMetricsClient {
+	return NewHeapsterMetricsClient(client, DefaultHeapsterNamespace, DefaultHeapsterScheme, DefaultHeapsterService, DefaultHeapsterPort)
+}
+
+func PodMetricsUrl(namespace string, name string) string {
+	return fmt.Sprintf("%s/namespaces/%s/pods/%s", MetricsRoot, namespace, name)
+}
+
+func NodeMetricsUrl(name string) string {
+	return fmt.Sprintf("%s/nodes/%s", MetricsRoot, name)
+}
+
+func (cli *HeapsterMetricsClient) GetNodeMetrics(nodeName string, selector string) ([]metrics_api.NodeMetrics, error) {
+	params := map[string]string{"labelSelector": selector}
+	resultRaw, err := GetHeapsterMetrics(cli, NodeMetricsUrl(nodeName), params)
+	if err != nil {
+		return []metrics_api.NodeMetrics{}, err
+	}
+	metrics := make([]metrics_api.NodeMetrics, 0)
+	if nodeName == "" {
+		err = json.Unmarshal(resultRaw, &metrics)
+		if err != nil {
+			return []metrics_api.NodeMetrics{}, fmt.Errorf("failed to unmarshall heapster response: %v", err)
+		}
+	} else {
+		var singleMetric metrics_api.NodeMetrics
+		err = json.Unmarshal(resultRaw, &singleMetric)
+		if err != nil {
+			return []metrics_api.NodeMetrics{}, fmt.Errorf("failed to unmarshall heapster response: %v", err)
+		}
+		metrics = append(metrics, singleMetric)
+	}
+	return metrics, nil
+}
+
+func (cli *HeapsterMetricsClient) GetPodMetrics(namespace string, podName string, allNamespaces bool, selector string) ([]metrics_api.PodMetrics, error) {
+	// TODO: extend Master Metrics API with getting pods from all namespaces
+	// instead of aggregating the results here
+	namespaces := make([]string, 0)
+	if allNamespaces {
+		list, err := cli.Client.Namespaces().List(api.ListOptions{})
+		if err != nil {
+			return []metrics_api.PodMetrics{}, err
+		}
+		for _, ns := range list.Items {
+			namespaces = append(namespaces, ns.Name)
+		}
+	} else {
+		namespaces = append(namespaces, namespace)
+	}
+
+	params := map[string]string{"labelSelector": selector}
+	allMetrics := make([]metrics_api.PodMetrics, 0)
+	for _, ns := range namespaces {
+		resultRaw, err := GetHeapsterMetrics(cli, PodMetricsUrl(ns, podName), params)
+		if err != nil {
+			return []metrics_api.PodMetrics{}, err
+		}
+		if podName == "" {
+			metrics := make([]metrics_api.PodMetrics, 0)
+			err = json.Unmarshal(resultRaw, &metrics)
```

Comment:
I'm not familiar with the metrics API.  This looks versioned (good), but we don't it through a scheme/codec style decoding?